packages/astro/src/core/build/plugins/plugin-pages.ts

```diff
@@ -36,31 +41,15 @@ function vitePluginPages(opts: StaticBuildOptions, internals: BuildInternals): V
 					i++;
 				}
 
-				i = 0;
-				let rendererItems = '';
-				for (const renderer of opts.settings.renderers) {
-					const variable = `_renderer${i}`;
-					// Use unshift so that renderers are imported before user code, in case they set globals
-					// that user code depends on.
-					imports.unshift(`import ${variable} from '${renderer.serverEntrypoint}';`);
-					rendererItems += `Object.assign(${JSON.stringify(renderer)}, { ssr: ${variable} }),`;
-					i++;
+				if (opts.settings.config.experimental.middleware) {
+					imports.push(`import * as _middleware from "${MIDDLEWARE_MODULE_ID}";`);
+					exports.push(`export const middleware = _middleware;`);
 				}
 
-				const def = `${imports.join('\n')}
-
-${
-	opts.settings.config.experimental.middleware
-		? `import * as _middleware from "${MIDDLEWARE_MODULE_ID}";`
-		: ''
-}
-
-export const pageMap = new Map([${importMap}]);
-export const renderers = [${rendererItems}];
-${opts.settings.config.experimental.middleware ? `export const middleware = _middleware;` : ''}
-`;
+				content.push(`export const pageMap = new Map([${importMap}]);`);
 
-				return def;
+				const result = [imports.join('\n'), content.join('\n'), exports.join('\n')];
+				return result.join('\n');
```

Comment:
Maybe we should directly construct the string here instead of creating a new array and joining it? e.g.
```suggestion
				return `\
${imports.join('\n')}
${content.join('\n')}
${exports.join('\n')}`
```

Made a [perf test](https://perf.link/#eyJpZCI6InMxajVtbDIxYWF4IiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGZvbyA9IFwiTG9yZW0gSXBzdW0gaXMgc2ltcGx5IGR1bW15IHRleHQgb2YgdGhlIHByaW50aW5nIGFuZCB0eXBlc2V0dGluZyBpbmR1c3RyeS4gTG9yZW0gSXBzdW0gaGFzIGJlZW4gdGhlIGluZHVzdHJ5J3Mgc3RhbmRhcmQgZHVtbXkgdGV4dFwiXG5jb25zdCBiYXIgPSBcIkxvcmVtIElwc3VtIGlzIHNpbXBseSBkdW1teSB0ZXh0IG9mIHRoZSBwcmludGluZyBhbmQgdHlwZXNldHRpbmcgaW5kdXN0cnkuIExvcmVtIElwc3VtIGhhcyBiZWVuIHRoZSBpbmR1c3RyeSdzIHN0YW5kYXJkIGR1bW15IHRleHRcIiIsInRlc3RzIjpbeyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImZvbyArICdcXG4nICsgYmFyIiwicnVucyI6WzQ1OTMwMDAsNTY4MDAwLDEyMDkwMDAsNjg4MDAwMCw2ODU2MDAwLDQ1NTYwMDAsNjk4MDAwLDI5NjMwMDAsMjk2MTAwMCw4NzgwMDAsOTA0MzAwMCwzMDY2MDAwLDczMDAwMDAsNjI3NzAwMCw3ODIyMDAwLDYyOTMwMDAsNjEyMjAwMCwxMTkxODAwMCwyNDkyMDAwLDY0NTAwMCw3MjU2MDAwLDQwMDIwMDAsNTc2MzAwMCw0MTMyMDAwLDMxOTIwMDAsNTIzNTAwMCw2NzkwMDAsNjE5MDAwMCwyNTI4MDAwLDUxMTgwMDAsMzAyODAwMCwxMDI5NzAwMCwzOTkzMDAwLDI3MDAwMDAsMTA0MjcwMDAsMjM4MjAwMCw3NDk3MDAwLDMzODIwMDAsOTA4NDAwMCwzNTM1MDAwLDY1MjAwMDAsNjgxODAwMCwxMjMzNTAwMCwxOTkxMDAwLDcxNTIwMDAsMzYwMTAwMCw3NTUyMDAwLDY4NTUwMDAsMTU2MzcwMDAsMTEyNzYwMDAsMjcyNTAwMCwxNjE0MjAwMCwxMzc1MzAwMCw2NDgwMDAwLDY1NjkwMDAsNTM4MTAwMCwyMjU2MDAwLDE0NTY1MDAwLDEwOTgwMDAsNjk2NDAwMCw2NTMzMDAwLDEzMDU2MDAwLDQ2NjAwMDAsMTY1NDAwMCwxMjI1NzAwMCw1NDEwMDAwLDkwMDkwMDAsNjc3OTAwMCwyMTUwMDAsMTU3NDMwMDAsNzQyMDAwMCwxNjg4NDAwMCw4OTY5MDAwLDY5NTAwMCwxMjY0OTAwMCw1OTMyMDAwLDE2MDYwMDAwLDg0ODkwMDAsNzg2MDAwLDExNTM1MDAwLDM1MDcwMDAsMTU3MTkwMDAsNDk3NDAwMCw4ODYwMDAwLDU4OTAwMDAsODc1MDAwLDQ5NTAwMCwxMDA2MTAwMCw1MDI3MDAwLDUwNjAwMCw5NzM3MDAwLDUwNzEwMDAsODIyMTAwMCw3MDI5MDAwLDk1MjEwMDAsMTI1MTAwMCwxMTc2ODAwMCw5OTkwMDAwLDcyNjUwMDAsNDIzNzAwMF0sIm9wcyI6NjQzOTY5MH0seyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImAke2Zvb31cXG4ke2Jhcn1gIiwicnVucyI6WzEwMTEwMDAsNDk4MDAwLDIxNzcwMDAsNzYzOTAwMCwxNjY2MDAwLDYxNDIwMDAsMTUwMjEwMDAsMjMyNzAwMCw0NjkzMDAwLDkzMjAwMCw3NTcxMDAwLDI1NDEwMDAsNjI0NjAwMCw0MjU1MDAwLDM5MDcwMDAsODU1ODAwMCwzNTIxMDAwLDgyNTIwMDAsNjg1OTAwMCw2NjMwMDAsNzU2MjAwMCw1MDI3MDAwLDYxMzAwMDAsMTUzNzAwMCwyMDg1MDAwLDEwMDYzMDAwLDEzMDYwMDAsNzU1ODAwMCwyNTU3MDAwLDUyNTkwMDAsNDAyMzAwMCw2NjczMDAwLDI5NDcwMDAsMTExNzAwMCw0MjMyMDAwLDM0MDYwMDAsOTE1MDAwMCw1MjEwMDAsOTg3MTAwMCw2NTU4MDAwLDUwMjgwMDAsNDAzNjAwMCw5NTIxMDAwLDMzMDUwMDAsMzg3MzAwMCwzODcyMDAwLDEyNTA1MDAwLDUyMjYwMDAsNDAzMDAwLDg5NTIwMDAsNTkzMDAwLDEwMjAwMCw1NDYzMDAwLDM0NjkwMDAsNjc0OTAwMCwyOTUwMDAwLDIyNjUwMDAsOTM3NTAwMCwxNTM2MDAwLDY5MzEwMDAsNDE5MTAwMCwxMzA1OTAwMCwzODAxMDAwLDI1NzcwMDAsNzQzNDAwMCw3MjUyMDAwLDEwNTE0MDAwLDY4MzYwMDAsMTEwNzAwMCwxNjA1NTAwMCw5Mjc5MDAwLDE5MjA2MDAwLDk1NDUwMDAsMTYzNzAwMCwxMTQ5ODAwMCw4MDM5MDAwLDE0NjE1MDAwLDc5NTEwMDAsNDEzNDAwMCwxMDUyMDAwMCw1NDgyMDAwLDE4MDQ0MDAwLDQwMzIwMDAsNzQ1NTAwMCw2MjMxMDAwLDI4NDgwMDAsOTY3MDAwLDk0OTcwMDAsNzE1NzAwMCw3MjAwMDAsODAxNDAwMCw2NjM4MDAwLDEwNDYxMDAwLDUyNjIwMDAsNjAzNDAwMCwxMzk0MDAwLDgzNDMwMDAsMTA0MTQwMDAsNzU0NTAwMCw0NjY3MDAwXSwib3BzIjo1ODg2NzAwfSx7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoiW2ZvbywgYmFyXS5qb2luKCdcXG4nKSIsInJ1bnMiOls2NjUwMDAsOTcwMDAwLDI4NzYwMDAsMjg3NjAwMCwyODc2MDAwLDIxMzIwMDAsNTE2MDAwLDE5NDgwMDAsMjQzOTAwMCw1MDIwMDAsMjg3NjAwMCwxNDczMDAwLDIzOTYwMDAsMjg3NjAwMCwyMTg3MDAwLDI4NzYwMDAsMTgyNDAwMCwyODc2MDAwLDEwNjUwMDAsNjYzMDAwLDI4NzYwMDAsMjIxMzAwMCwyODc2MDAwLDE3NjMwMDAsMjAxODAwMCwyODc2MDAwLDU2NzAwMCwyODc2MDAwLDE4MTUwMDAsMjM2NTAwMCwyMjA4MDAwLDI4NzYwMDAsMjE5MTAwMCw1NzQwMDAsMjg3NjAwMCwxNzc0MDAwLDI4NzYwMDAsODcwMDAwLDI4NzYwMDAsMjcyNzAwMCwyMzkyMDAwLDIxNzQwMDAsMjg3NjAwMCw5NTkwMDAsMjg3NjAwMCwyMzcxMDAwLDI4NzYwMDAsMTY1MTAwMCwyOTkxMDAwLDI4NzYwMDAsMTcwNzAwMCwyOTAwMDAwLDI1ODUwMDAsMTkwMDAwMCwyODc2MDAwLDY2NzAwMCwyNTcwMDAsMjg3NjAwMCw2OTIwMDAsMjg3NjAwMCwyODc2MDAwLDI4NzYwMDAsMjg3NjAwMCwxMDA2MDAwLDI4NzYwMDAsMjg3NjAwMCwyODc2MDAwLDI4NzYwMDAsMTAwMCwyODc2MDAwLDI4NzYwMDAsMTIxMDAwLDI4NzYwMDAsODI4MDAwLDI4NzYwMDAsMTUyNDAwMCwyODkzMDAwLDI4NzYwMDAsMjI4NzAwMCwyODc2MDAwLDI4NzYwMDAsMzEyNDAwMCwyNjQ1MDAwLDI4NzYwMDAsMjg3NjAwMCwyMzYzMDAwLDI1NzAwMCwyODc2MDAwLDI4NzYwMDAsNTcyMDAwLDI4NzYwMDAsOTY5MDAwLDI4NzYwMDAsMjY0MTAwMCwyODc2MDAwLDc4MjAwMCwyODc2MDAwLDI4NzcwMDAsMjg3NjAwMCwyNjI1MDAwXSwib3BzIjoyMjAwNDYwfV0sInVwZGF0ZWQiOiIyMDIzLTA1LTIzVDA4OjM2OjQyLjk1M1oifQ%3D%3D) here too

Reply:
The main reason I used a temporary variable is debugging purpose only - so I can log the result code.

I can remove if you think this is an hot path

Reply:
I don't think it's a hot path, but I usually lean towards writing performant code whenever possible.